src/formatter.rs

```diff
@@ -10,44 +10,113 @@ lazy_static! {
     static ref RE_CONSTANT: Regex = Regex::new(r"\{\{((?s).*?)\}\}").unwrap();
     static ref RE_BOLD: Regex = Regex::new(r"<<((?s).*?)>>").unwrap();
     static ref RE_MONOSPACE: Regex = Regex::new(r"`([^`]*?)`").unwrap();
-    static ref RE_MONOSPACE_TRIM: Regex = Regex::new(r"\n? *(`[^`]*`) *").unwrap();
+    static ref RE_MONOSPACE_OLD: Regex = Regex::new(r"```([^`]*?)```").unwrap();
+    static ref RE_MONOSPACE_TRIM: Regex = Regex::new(r"\s*`\s*([^`]+?)\s*`\s*").unwrap();
     static ref RE_BACKTICK: Regex = Regex::new(r"(`[^`]+`)|([^`]+)").unwrap();
     static ref RE_SPACES: Regex = Regex::new(r" +").unwrap();
     static ref RE_NONWHITESPACE: Regex = Regex::new(r"[^\r\n ]+").unwrap();
+    static ref RE_NEWLINES: Regex = Regex::new(r"\n\n\n+").unwrap();
+    static ref RE_ALL_BUT_MONOSPACE: Regex =
+        Regex::new(r"\[\[((?s).*?)\]\]|\{\{((?s).*?)\}\}|<<((?s).*?)>>").unwrap();
 }
 
-/// Format Codingame statement that contains special formatting syntax
-/// [[VARIABLE]] - {{CONSTANT}} - <<BOLD>> - `MONOSPACE`
 pub fn format_cg(text: &str, ostyle: &OutputStyle) -> String {
-    // Trim consecutive spaces (imitates html behaviour)
-    // But only if it's not in a Monospace block (between backticks ``)
-    let mut result = RE_BACKTICK
-        .replace_all(text, |caps: &regex::Captures| {
-            if let Some(backtick_text) = caps.get(1) {
-                backtick_text.as_str().to_string()
-            } else if let Some(non_backtick_text) = caps.get(2) {
-                RE_SPACES.replace_all(non_backtick_text.as_str(), " ").to_string()
+    let mut text = format_edit_monospace(&text, ostyle);
+    text = format_trim_consecutive_spaces(&text);
+    text = format_monospace_padding(&text);
+    text = format_add_reverse_nester_tags(&text);
+    text = format_paint_inner_blocks(&text, ostyle);
+    format_remove_excessive_newlines(&text)
+}
+
+/// 1. Throws a warning if it finds outdated formatting
+/// 2. Replaces ```text``` -> `text`
+/// 3. Format whitespace around Monospace blocks.
+///
+/// Clashes with outdated formatting:
+///     https://www.codingame.com/contribute/view/25623694f80d8f747b3fa474a33a9920335ce
+///     https://www.codingame.com/contribute/view/7018d709bf39dcccec4ed9f97fb18105f64c
+/// Others:
+///     https://www.codingame.com/contribute/view/1222536cec20519e1a630ecc8ada367dd708b
+fn format_edit_monospace(text: &str, ostyle: &OutputStyle) -> String {
+    if RE_MONOSPACE_OLD.is_match(&text) {
+        let msg = "Clash contains obsolete ``` formatting, consider fixing it in the website.";
+        println!("{} {}\n", ostyle.failure.paint("WARNING"), msg);
+    }
+
+    let mut result = text.replace("```", "`");
+
+    result = RE_MONOSPACE_TRIM
+        .replace_all(&result, |caps: &regex::Captures| format!("\n\n`{}`\n\n", &caps[1]))
```

Comment:
Shouldn't this only be one '\n' when the string doesn't have any newlines around the monospace? For example I would expect
```
abc `def` ghi
```
to become
```
abc
def
ghi
```
(with formatting)

Reply:
This is something where I think not following the CG logic turns out to be prettier in the terminal. I did implement an exact copy of the Monospace formatting that CG does... and it looked worse, with an extra 50+ lines of very hard to read code.

I find it a byproduct of the reasons why people normally use Monospace in codingame, which is to separate a block of text (and less about spacing), and the fact that CG renders newlines differently if they are inside a Monospace block or not, which is not reproducible in the terminal.

Have you tested how it looks changing that \n\n into \n for the links around the file? Do you still prefer only one?

Reply:
There's a lot of clashes that use monospace inside a sentence and I think those ones can look a bit a weird with the blank lines around them. Here's an example:

`cargo run -- show 7720e12fdb9f17545450dec43204fec9a172`
![image](https://github.com/Andriamanitra/clash/assets/10672443/f1a78915-ae86-4e4c-b028-c27f2026c207)

![image](https://github.com/Andriamanitra/clash/assets/10672443/fdbe40db-5cb8-410f-aa21-2c2a2a9e13a3)

I slightly prefer the one without the blank lines in this case, but I'm not sure if the improvement is big enough to be worth the added complexity of checking if there were newlines around the backtick.